Implement deletePost

mccodav1 opened the pull request:
This commit implements post deletion from the client.

- a `deletePost` method was added to `MyApiService` which defines the route DELETE:"posts/" to make an API call to delete the Post from the backend

- `deletePostFromServer` method added to `com.example.poste.models.User` class, which makes the API call to the server to delete a Post. It takes a Post model (defined in `com.example.poste.models.Post) and a callback as parameters. Callback will be discussed below.

- `PostDeletionCallback` was added to `com.example.poste.calbacks`. Please review the code for usage details; I've added extensive docstrings. Since API calls are made asynchronously, we cannot depend on return values like we might in a standard Java application. A 'callback' informs the calling code of asynchronous success or failure -- this way, we can decide what to do if a request succeeds or fails without waiting on the response.

- `DebugUtils` has had some extra utilities added

- RetrofitClient was slightly modified to add a logger

TL;DR: To add post deletion to an activity:
1) Define a PostDeletionCallback in your activity, implementing `onSuccess` and `onError` methods. 2) Call `User.getUser().deletePostFromServer`, passing a Post model and the callback. 3) If the call succeeds, the `onSuccess` method will run. Otherwise, the `onError` method will run. 4) If you have questions, look at the `User.deletePostFromServer` docstring

mccodav1 committed:
Implement deletePost

This commit implements post deletion from the client.

- a `deletePost` method was added to `MyApiService` which defines the route DELETE:"posts/"
to make an API call to delete the Post from the backend

- `deletePostFromServer` method added to `com.example.poste.models.User` class, which makes the
API call to the server to delete a Post. It takes a Post model (defined in
`com.example.poste.models.Post) and a callback as parameters. Callback will be discussed below.

- `PostDeletionCallback` was added to `com.example.poste.calbacks`. Please review the code for
usage details; I've added extensive docstrings. Since API calls are made asynchronously, we cannot
depend on return values like we might in a standard Java application. A 'callback' informs the
calling code of asynchronous success or failure -- this way, we can decide what to do if a request
succeeds or fails without waiting on the response.

- `DebugUtils` has had some extra utilities added

- RetrofitClient was slightly modified to add a logger

TL;DR: To add post deletion to an activity:
1) Define a PostDeletionCallback in your activity, implementing `onSuccess` and `onError` methods.
2) Call `User.getUser().deletePostFromServer`, passing a Post model and the callback.
3) If the call succeeds, the `onSuccess` method will run. Otherwise, the `onError` method will run.
4) If you have questions, look at the `User.deletePostFromServer` docstring

mccodav1 commented on app/src/main/java/com/example/poste/http/MyApiService.java:
```diff
@@ -28,4 +29,32 @@ public interface MyApiService {
+    @DELETE("posts/")
+    Call<ResponseBody> deletePost(
+            @Header("Authorization") String authToken,
+            @Header("id") String id
```
I don't know why I did it this way. After thinking on it, it makes more sense to use id in the URL path rather than in a header or body. I have a branch to fix this, pending update tomorrow, as I need to also update the backend.

atreichl replied:
I agree with the ID being in the URL path.

mccodav1 commented on app/src/main/java/com/example/poste/http/RetrofitClient.java:
```diff
@@ -10,7 +11,11 @@ public class RetrofitClient {
+            HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
+            logging.setLevel(HttpLoggingInterceptor.Level.HEADERS);
+            OkHttpClient client = new OkHttpClient.Builder()
+                    .addInterceptor(logging)
```
This adds logging to our HTTP request client, for debugging purposes.

atreichl replied:
This is a good idea.

mccodav1 commented on app/src/main/java/com/example/poste/models/User.java:
```diff
@@ -246,7 +247,100 @@ public void onFailure(Call<ResponseBody> call, Throwable t) {
+     * @param post The Post model to delete.
+     * @param callback The callback to be invoked after attempting to delete the post.
+     */
+    public void deletePostFromServer(Post post, PostDeletionCallback callback) {
```
In consideration, we should probably use callbacks in more places.

I found this after googling why I wasn't having success returning booleans from functions, and it has to do with the asynchronous calls happening inside.

A callback effectively allows you to specify what you'll do if an async event (such as an API call) succeeds or fails (really, you can define any number of situations). When you make the request, you can make certain callbacks to effect change.

mccodav1 commented on app/src/main/java/com/example/poste/models/User.java:
```diff
@@ -246,7 +247,100 @@ public void onFailure(Call<ResponseBody> call, Throwable t) {
+        Log.d("UserDebug", "Sending delete request to API for Post ID " + id);
+        Call<ResponseBody> call = apiService.deletePost(getTokenHeader(), id);
+
+        call.enqueue(new Callback<ResponseBody>() {
```
Here, we see we've actually been using Callbacks all along, without much consideration to their meaning

atreichl reviewed (changes requested):
It looks good to me, I'll approve it once you change the id to part of the URL path.

mccodav1 committed:
Fix delete posts path and minor bugfix

mccodav1 committed:
Merge branch 'dev' of github.com:Appyo-Poste/Poste into us68_t107

Resolves merge conflicts from pulling dev changes

mccodav1 commented:
> It looks good to me, I'll approve it once you change the id to part of the URL path.

Done.

atreichl commented on app/src/main/java/com/example/poste/activities/LoginActivity.java:
```diff
@@ -109,7 +109,6 @@ public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response)
                             Log.e("Json Error", err.toString());
                         }
                         // open dashboard activity
-                        User.getUser().updateFoldersAndPosts(LoginActivity.this);
```
This is probably caused by newer merges happening since you last merged dev in, but if this line is removed dashboard will not have any folders when it launches.

mccodav1 replied:
I guess I assumed this was occurring in DashboardActivity after a discussion I had with Fayez, but I'll add it back here.

atreichl reviewed (changes requested):
fix the one minor error I mentioned and it's good to go.

mccodav1 committed:
Resolve dev merge conflicts and refactor

- This commit resolves merge conflicts from dev and fixes a myriad of
conflicts / styling issues / concerns. It attempts to integrate some of
the activities / flows that were refactored by different developers.

mccodav1 commented on app/src/main/java/com/example/poste/activities/EditPostActivity.java:
```diff
@@ -49,47 +51,57 @@ protected void onCreate(Bundle savedInstanceState) {
         DebugUtils.logUserFoldersAndPosts(User.getUser());
 
-        EditText linkToPost = findViewById(R.id.editTextLinkToPost);
+        EditText postLink = findViewById(R.id.editTextPostLink);
```
Just to keep naming conventions alike

mccodav1 commented on app/src/main/java/com/example/poste/http/DeletePost.java:
Similar to EditPost or DeleteFolder, we'll want to pass in just a String here (for path ID) rather than creating an object (nothing goes in body)

mccodav1 commented on app/src/main/java/com/example/poste/http/utils.java:
Extracted this from User into its own static method

mccodav1 commented on app/src/main/java/com/example/poste/models/Folder.java:
```diff
@@ -82,12 +81,11 @@ public String getUserPermission() {
-     * Returns a copy of the list of posts in the folder. This is a copy, so modifying the returned
-     * list will not modify the folder's posts.
-     * @return a copy of the list of posts in the folder.
+     * Returns a List of the posts in the folder.
```
I like the idea of immutable return values here (returning a copy, instead of the actual list) -- however, I think this was causing some confusion and we were attempting to modify the returned lists of folders / posts. Until we can get on the same page and refactor, this will solve some issues.

mccodav1 commented on app/src/main/java/com/example/poste/models/User.java:
```diff
@@ -177,17 +176,17 @@ public Post getPost(String postId) {
      */
     public List<Folder> getFolders() {
-        return new ArrayList<>(folders);
+        return folders;
```
I like the idea of immutable return values here (returning a copy, instead of the actual list) -- however, I think this was causing some confusion and we were attempting to modify the returned lists of folders / posts. Until we can get on the same page and refactor, this will solve some issues.

dzemlin replied:
Fair enough

mccodav1 commented on app/src/main/java/com/example/poste/models/User.java:
```diff
@@ -202,7 +201,7 @@ public String getToken() {
      * @return the user's token, prepended with "Token "
      */
-    public String getTokenHeader() {
+    public String getTokenHeaderString() {
```
Refactored name for clarity

atreichl reviewed (approved):
Everything looks good to me now.

dzemlin commented on build.gradle:
```diff
@@ -1,6 +1,6 @@
-    id 'com.android.library' version '8.1.1' apply false
+    id 'com.android.application' version '8.0.2' apply false
+    id 'com.android.library' version '8.0.2' apply false
     id 'org.jetbrains.kotlin.android' version '1.8.10' apply false
```
shouldn't we leave the build version at 8.1.1 since it forces us to update to that every time we run?

mccodav1 replied:
I have the opposite problem and I imagine might be something local to your install; the project has historically been on 8.0.2. I can change it to 8.1.1 but that will just cause the reverse issues for others...

dzemlin commented on app/src/main/java/com/example/poste/models/User.java:
```diff
@@ -106,46 +131,20 @@ public void setLastName(String lastName) {
      */
-    public void addFolder(Folder newFolder) {
+    private void addFolder(Folder newFolder) {
         folders.add(newFolder);
```
if this causse merge conflicts with my debug data model method in PosteApplicatin.java, feel free to comment out that method.

mccodav1 replied:
I think we might be using the term 'merge conflict' differently; can you elaborate what you mean?

dzemlin reviewed (approved):
Looks good to me. but I think a third set of eyes would be a good idea too.

mccodav1 committed:
Upgrade build.gradle android vers
